Switch from web3.js to ethers.js (#15760)
ethereum/solidity: docs/introduction-to-smart-contracts.rst

Reviewer: Note that this is not the only place where we're using web3.js. We have a few more examples/mentions:
- [Events](https://docs.soliditylang.org/en/develop/contracts.html#events)
- [Micropayment channel example](https://docs.soliditylang.org/en/develop/solidity-by-example.html#micropayment-channel)
- [Creating contracts](https://docs.soliditylang.org/en/develop/contracts.html#creating-contracts)

We should probably also mention deprecation in the "Resources" section.

Reviewer 2: I think there's also a couple of mentions/examples in [Solidity by Example](https://docs.soliditylang.org/en/v0.8.29/solidity-by-example.html) that could be adjusted.
@codersjj , are you still interested in working this?
Sorry for the long time to come back to this. Hopefully we can merge it soon now.

transactions.

To listen for this event, you could use the following

> JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to create the ``coin`` contract object,

Reviewer 2: Although it was already phrased like this before, maybe we could do something regarding that the text says the code "uses ether.js to **create** the coin contract object ..." but there is no line actually creating any object.
Maybe we should just change to:
```suggestion
JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to listen to ``Sent`` events broadcasted from the ``Coin`` contract ,
```

Reviewer 2: Alternatively, we could also add a "templated" line of code which creates a ``Coin`` contract object:

```javascript
const coin = new ethers.Contract(coinAddress, abi, provider);
```

Reviewer 2: Does that make sense, @r0qs ?

> JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to create the ``coin`` contract object,

Reviewer 2: ```suggestion
JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to create the ``Coin`` contract object,
```
Here `Coin` refers to the contract name, not the variable name in the example.

Author: > Here `Coin` refers to the contract name, not the variable name in the example.

Got it, thanks for clarifying!

> JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to listen to ``Sent`` events broadcasted from the ``Coin`` contract ,

Reviewer: ```suggestion
JavaScript code, which uses `ethers.js <https://github.com/ethers-io/ethers.js>`_ to listen to ``Sent`` events broadcasted from the ``Coin`` contract,
```

and any user interface calls the automatically generated ``balances`` function from above:

.. code-block:: javascript

    coin.on('Sent', async (from, to, amount, event) => {
      console.log(`Coin transfer: ${ethers.formatUnits(amount, 18)} coins were sent from ${from} to ${to}.`);
    

>       const senderBalance = await contract.balances(from);
>       const receiverBalance = await contract.balances(to);

Reviewer 2: ```suggestion
      const senderBalance = await coin.balances(from);
      const receiverBalance = await coin.balances(to);
```

I gues this should be ``coin`` as that is the name used before to listen to the event, right?

Author: > I gues this should be `coin` as that is the name used before to listen to the event, right?

Good catch, thanks!

>       console.log(`Balances now:\nSender: ${ethers.formatUnits(senderBalance, 18)}\nReceiver: ${ethers.formatUnits(receiverBalance, 18)}`);

Reviewer: The original formatting here was more readable. No reason to remove it.
```suggestion
        console.log(
            `Balances now:\n` +
            `Sender: ${ethers.formatUnits(senderBalance, 18)}\n` +
            `Receiver: ${ethers.formatUnits(receiverBalance, 18)}`
        );
```
